### PR TITLE
created tests for fruit shop

### PR DESCRIPTION
**Pull request by Anastasia-702**

### CODE REVIEW

File: src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class FruitTransactionServiceImplTest {
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java:23]:
```suggestion
public class FruitTransactionServiceTest {
```
the same for other similar cases

File: src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        fruitTransactionService = new FruitTransactionServiceImpl(strategy,fruitsStorage);
+    }
+
+    @Before
+    public void setUp() {
+        strategyMap.put(FruitTransaction.Operation.BALANCE, new BalanceOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.PURCHASE,
+                new PurchaseOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.RETURN, new ReturnOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.SUPPLY, new SupplyOperationHandler());
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java:45]: let's move this logic to beforeClass method

File: src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Test
+    public void processTransaction_Ok() {
+        FruitTransaction fruitTransaction = new FruitTransaction();
+        fruitTransaction.setOperation(FruitTransaction.Operation.BALANCE);
+        fruitTransaction.setFruit("banana");
+        fruitTransaction.setQuantity(20);
+        transactions.add(fruitTransaction);
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/FruitTransactionServiceImplTest.java:54]: you use the same code several times, let's eliminate duplication and create a separate method that will be creating FruitTransaction based on passed parameters

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        dataFromFile = new ArrayList<>();
+    }
+
+    @Test
+    public void getTransactionFromString_Ok() {
+        dataFromFile.add("b,banana,20");
+        dataFromFile.add("b,apple,100");
+        dataFromFile.add("s,banana,100");
+        dataFromFile.add("p,banana,13");
+        final List<FruitTransaction> expectedResult = new ArrayList<>();
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:28]:
```suggestion
        List<FruitTransaction> expectedResult = new ArrayList<>();
```

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        FruitTransaction fruitTransaction = new FruitTransaction();
+        fruitTransaction.setOperation(FruitTransaction.Operation.BALANCE);
+        fruitTransaction.setFruit("banana");
+        fruitTransaction.setQuantity(20);
+        expectedResult.add(fruitTransaction);
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:33]: you use the same code several times, let's eliminate duplication and create a separate method that will be creating FruitTransaction based on passed parameters

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        fruitTransaction3.setFruit("banana");
+        fruitTransaction3.setQuantity(13);
+        expectedResult.add(fruitTransaction3);
+        List<FruitTransaction> actualResult = parserService.getTransactionFromString(dataFromFile);
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void getTransactionFromString_emptyInput_notOk() {
+        parserService.getTransactionFromString(dataFromFile);
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:55]:
```suggestion
        parserService.parse(Collections.emptyList());
```

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class ParserServiceImplTest {
+    private static ParserService parserService;
+    private static List<String> dataFromFile;
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:14]: let's make it local

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.List;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class WriterServiceImplTest {
+    private static WriterService writerService;
+    private static List<String> report;
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:18]: let's make it local

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        List<String> actualResult = readFromFile("src/test/resources/report.csv");
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void writeToFile_directoryDoesNotExists_notOk() {
+        writerService.writeToFile(report, "src/test/nonexistent/report.csv");
+    }
+
+    public List<String> readFromFile(String filePath) {
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:45]: do you use this method outside this class?

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    public void writeToFile_Ok() {
+        writerService.writeToFile(report, "src/test/resources/report.csv");
+        List<String> expectedResult = List.of("fruit, quantity", "banana, 100");
+        List<String> actualResult = readFromFile("src/test/resources/report.csv");
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void writeToFile_directoryDoesNotExists_notOk() {
+        writerService.writeToFile(report, "src/test/nonexistent/report.csv");
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:42]:
```suggestion
        writerService.writeToFile(Collections.emptyList(), "src/test/nonexistent/report.csv");
```

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+            dataFromFile = Files.readAllLines(Path.of(filePath));
+        } catch (IOException e) {
+            throw new RuntimeException("Can't read data from file " + filePath, e);
+        }
+        return dataFromFile;
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        new FileWriter("src/test/resources/report.csv").close();
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:57]: Do you need this method?

File: src/test/java/core/basesyntax/strategy/BalanceOperationHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        fruitsStorage = new Storage();
+        transaction = new FruitTransaction();
+    }
+
+    @Test
+    public void processOperation_Ok() {
+        transaction.setFruit("banana");
+        transaction.setQuantity(22);
+        balanceOperationHandler.processOperation(transaction, fruitsStorage);
+        int actual = fruitsStorage.getFruitsStorage().get("banana");
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/BalanceOperationHandlerTest.java:27]:
```suggestion
        int actual = fruitsStorage.getFruitsStorage().get(transaction.getFruit());
```

File: src/test/java/core/basesyntax/strategy/FruitTransactionStrategyImplTest.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        fruitTransactionStrategy = new FruitTransactionStrategyImpl(strategyMap);
+    }
+
+    @Before
+    public void setUp() {
+        strategyMap.put(FruitTransaction.Operation.BALANCE, new BalanceOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.PURCHASE,
+                new PurchaseOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.RETURN, new ReturnOperationHandler());
+        strategyMap.put(FruitTransaction.Operation.SUPPLY, new SupplyOperationHandler());
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/FruitTransactionStrategyImplTest.java:27]: move this logic to beforeClass() method

File: src/test/java/core/basesyntax/strategy/FruitTransactionStrategyImplTest.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class FruitTransactionStrategyImplTest {
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/FruitTransactionStrategyImplTest.java:11]: let's add a test for the case if null is passed to get method

File: src/test/java/core/basesyntax/strategy/PurchaseOperationHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        transaction = new FruitTransaction();
+    }
+
+    @Test
+    public void processOperation_Ok() {
+        fruitsStorage.getFruitsStorage().put("banana", 100);
+        transaction.setFruit("banana");
+        transaction.setQuantity(90);
+        purchaseOperationHandler.processOperation(transaction, fruitsStorage);
+        int actual = fruitsStorage.getFruitsStorage().get("banana");
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/PurchaseOperationHandlerTest.java:28]:
```suggestion
        int actual = fruitsStorage.getFruitsStorage().get(transaction.getFruit());
```

File: src/test/java/core/basesyntax/strategy/ReturnOperationHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        transaction = new FruitTransaction();
+    }
+
+    @Test
+    public void processOperation_Ok() {
+        fruitsStorage.getFruitsStorage().put("banana", 1);
+        transaction.setFruit("banana");
+        transaction.setQuantity(22);
+        returnOperationHandler.processOperation(transaction, fruitsStorage);
+        int actual = fruitsStorage.getFruitsStorage().get("banana");
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/ReturnOperationHandlerTest.java:28]:
```suggestion
        int actual = fruitsStorage.getFruitsStorage().get(transaction.getFruit());
```

File: src/test/java/core/basesyntax/strategy/SupplyOperationHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        transaction = new FruitTransaction();
+    }
+
+    @Test
+    public void processOperation_Ok() {
+        fruitsStorage.getFruitsStorage().put("banana", 10);
+        transaction.setFruit("banana");
+        transaction.setQuantity(22);
+        supplyOperationHandler.processOperation(transaction, fruitsStorage);
+        int actual = fruitsStorage.getFruitsStorage().get("banana");
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/strategy/SupplyOperationHandlerTest.java:28]:
```suggestion
        int actual = fruitsStorage.getFruitsStorage().get(transaction.getFruit());
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/impl/FruitTransactionServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        Map<String, Integer> actualResult = fruitsStorage.getFruitsStorage();
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void processTransaction_emptyInput_notOk() {
+        fruitTransactionService.processTransaction(Collections.emptyList());
+    }
+
+    private FruitTransaction createFruitTransaction(String fruit, int quantity,
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/FruitTransactionServiceTest.java:82]: private methods should go after public ones

File: src/test/java/core/basesyntax/service/impl/FruitTransactionServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+        fruitTransactionService = new FruitTransactionServiceImpl(strategy,fruitsStorage);
+    }
+
+    @Test
+    public void processTransaction_Ok() {
+        List<FruitTransaction> transactions = new ArrayList<>();
+        FruitTransaction fruitTransaction = createFruitTransaction("banana", 20,
+                FruitTransaction.Operation.BALANCE);
+        transactions.add(fruitTransaction);
+        FruitTransaction fruitTransaction1 = createFruitTransaction("apple", 100,
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/FruitTransactionServiceTest.java:48]:
```suggestion
        FruitTransaction firstFruitTransaction = createFruitTransaction("apple", 100,
```
the same for other similar variables

File: src/test/java/core/basesyntax/service/impl/ParserServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        FruitTransaction fruitTransaction = new FruitTransaction();
+        fruitTransaction.setFruit(fruit);
+        fruitTransaction.setQuantity(quantity);
+        fruitTransaction.setOperation(operation);
+        return fruitTransaction;
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void getTransactionFromString_emptyInput_notOk() {
+        parserService.getTransactionFromString(Collections.emptyList());
```
Comment (Nick97-git) [inline src/test/java/core/basesyntax/service/impl/ParserServiceTest.java:55]:
```suggestion
        parserService.parse(Collections.emptyList());
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

**Review by Nick97-git — APPROVED** [review]
